### PR TITLE
Accessibility

### PR DESCRIPTION
**Pull request by jquense**
Some of this is just hard to do without a good answer to: https://github.com/facebook/react/issues/4000 

I've posted a comment asking if there was anything that was feasible for libraries to do
https://github.com/facebook/react/issues/4000#issuecomment-113044968

### CODE REVIEW

File: src/NavItem.js [hunk 1]
```diff
@@ -23,32 +26,37 @@ const NavItem = React.createClass({
...
-    if (href === '#') {
+    if (!role && href === '#') {
       linkProps.role = 'button';
     }
 
     return (
-      <li {...props} className={classNames(props.className, classes)}>
-        <a {...linkProps}>
+      <li {...props} role='presentation' className={classNames(props.className, classes)}>
+        <a {...linkProps} aria-selected={active} aria-controls={ariaControls}>
```
Comment (AlexKVal) [inline src/NavItem.js]: Isn't it supposed to be as `ariaSelected` and `ariaControls` ?
https://facebook.github.io/react/docs/tags-and-attributes.html#supported-attributes
> Note:
All attributes are camel-cased

Comment (jquense) [inline src/NavItem.js]: Yup with the exception of data and aria attributes :) not sure why that is

> React supports all data-* and aria-* attributes as well as every attribute in the following lists.

Comment (AlexKVal) [inline src/NavItem.js]: Yep. I've checked it now... 
It has to be like `aria-*`. So pity.

**Comment by AlexKVal** [comment]
Cool :sparkle:

File: src/NavItem.js [hunk 2]
```diff
@@ -23,32 +26,37 @@ const NavItem = React.createClass({
...
     let {
+        role,
+        linkId,
         disabled,
         active,
         href,
         title,
         target,
         children,
+        'aria-controls': ariaControls,  // eslint-disable-line react/prop-types
```
Comment (AlexKVal) [inline src/NavItem.js]: I created issue
https://github.com/yannickcr/eslint-plugin-react/issues/118

Comment (jquense) [inline src/NavItem.js]: :+1: its interesting that it isn't supported

**Comment by jquense** [comment]
anyone know why Appveyor is failing?

**Comment by mathieumg** [comment]
npm connectivity issues it would seem?

**Comment by mtscout6** [comment]
I re-triggered AppVeyor, npm installs have struggled to be reliable on AppVeyor for some reason...

**Comment by jquense** [comment]
AppVeyor hates my PRs :P

**Comment by dozoisch** [comment]
a11y for the win! Nice job

**Comment by jquense** [comment]
Does anyone have any suggestion about how we proceed with the ID issue? @mtscout6 did you make any progress in your Dropdown work? I saw that was a task in the issue.

It seems like we practically have about 3 options
- Generate ID's naively and tell users to supply their own id's when server rendering (so they match)
- only add the a11y stuff when a user supplies an id, (like [this](https://github.com/react-bootstrap/react-bootstrap/pull/857/files#diff-a8a38bd1bdd1a56c3f27491f45d5b8e3R8))
- require the user to wrap there App component in a utility we provide like: https://github.com/facebook/react/issues/4000#issuecomment-113097875

Ideally I'd like to add a11y stuff by default and have the user turn it off if necessary vs an opt in.

**Comment by mtscout6** [comment]
I was using [uuid](https://www.npmjs.com/package/uuid) to generate id's but once I saw facebook's issue I ripped it out primarily out of fear of delaying the dropdown any further and just made `id` a required prop.

**Comment by jquense** [comment]
haaaa, I like just making it a required prop. nice and simple. I added it to the list of options above

**Comment by mtscout6** [comment]
What I like about it being a required prop is that once facebook figures out a good solution we can remove the required prop. It would not break backwards compatibility since I see it as a good thing to allow users to customize if they wish.

**Comment by jquense** [comment]
ha finally green. Do folks want to merge this as is, and work on stuff that requires ID's in another PR against 0.24 or just add it all in here?

**Comment by mtscout6** [comment]
Id's should be fine here since required props are just warnings right?

**Comment by jquense** [comment]
true...its fine as long as the components don't break assuming its there b/c its required, but I can make sure that's true.

**Comment by mtscout6** [comment]
I guess if internally you break because it doesn't exist than it would make sense to go in v0.24. Maybe it's just safest to put in v0.24...

File: src/NavItem.js (hunk 2 above)
Comment (AlexKVal) [inline src/NavItem.js]: @Cellule has fixed it https://github.com/yannickcr/eslint-plugin-react/pull/124 :sparkle:

**Comment by mtscout6** [comment]
@jquense Were you planning to put the id stuff here or in another PR? If another PR we can get this one merged right?

**Comment by jquense** [comment]
yeah I think let's just merge this and I will work on another PR. these components are done as is anyway.  might as well get some improvements in now

**Pull request merged**